source/s3_endpoint.c

```diff
@@ -267,9 +299,16 @@ static void s_s3_endpoint_release(struct aws_s3_endpoint *endpoint) {
     /* BEGIN CRITICAL SECTION */
     aws_s3_client_lock_synced_data(endpoint->client);
 
-    bool should_destroy = (endpoint->client_synced_data.ref_count == 1);
+    bool should_destroy = endpoint->client_synced_data.ref_count == 1;
+    bool client_active = endpoint->client->synced_data.active == 1;
     if (should_destroy) {
-        aws_hash_table_remove(&endpoint->client->synced_data.endpoints, endpoint->host_name, NULL, NULL);
+        if (client_active) {
+            endpoint->client_synced_data.state = AWS_S3_ENDPOINT_STATE_PENDING_CLEANUP_TASK;
+            endpoint->client->synced_data.process_endpoint_lifecycle_changes = true;
+        } else {
+            endpoint->client_synced_data.state = AWS_S3_ENDPOINT_STATE_DESTROYING;
+            aws_hash_table_remove(&endpoint->client->synced_data.endpoints, endpoint->host_name, NULL, NULL);
```

Comment:
Question/Confusion: Is it worth avoiding aws_event_loop_schedule_task_now here? If we simplify the code here to always trigger aws_s3_client_schedule_process_work, we will clean up the endpoint on the client process work thread, but does it matter?